src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -6,47 +6,187 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Linq;
+using System.Runtime;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Operations;
 using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.Optimizations;
+using Microsoft.Quantum.QsCompiler.SymbolManagement;
 using Microsoft.Quantum.QsCompiler.SyntaxTokens;
 using Microsoft.Quantum.QsCompiler.SyntaxTree;
 using Microsoft.Quantum.QsCompiler.Transformations.Core;
 
 namespace Microsoft.Quantum.QsCompiler.Transformations.Monomorphization
 {
+    using Concretion = HashSet<(QsTypeParameter, ResolvedType)>;
+
+    internal struct Request
+    {
+        public QsQualifiedName originalName;
+        public Concretion typeResolutions;
+        public QsQualifiedName concreteName;
+    }
+
+    internal struct Response
+    {
+        public QsQualifiedName originalName;
+        public Concretion typeResolutions;
+        public QsCallable concreteCallable;
+    }
+
+    #region ResolveGenerics
+
     public class ResolveGenericsSyntax :
-        SyntaxTreeTransformation<ScopeTransformation<ResolveGenericsExpression>>
+        SyntaxTreeTransformation<NoScopeTransformations>
     {
-        public static IEnumerable<QsNamespace> Apply(IEnumerable<QsNamespace> namespaces)
+        ImmutableDictionary<NonNullable<string>, IEnumerable<QsCallable>> NamespaceCallables;
+
+        public static IEnumerable<QsNamespace> Apply(IEnumerable<QsNamespace> namespaces, QsQualifiedName entryPointName)
         {
             if (namespaces == null || namespaces.Contains(null)) throw new ArgumentNullException(nameof(namespaces));
 
-            // Get list of generics
-            List<QsQualifiedName> generics = new List<QsQualifiedName>();
-            var Callables = namespaces.GlobalCallableResolutions();
-            foreach (var callable in Callables)
+            var globals = namespaces.GlobalCallableResolutions();
+
+            var entryPoints = globals
+                .Where(call => // TODO: get list of entry points
+                    call.Key.Namespace.Value == entryPointName.Namespace.Value &&
+                    call.Key.Name.Value == entryPointName.Name.Value)
+                .Select(call => new Request
+                {
+                    originalName = call.Key,
+                    typeResolutions = null,
+                    concreteName = call.Key
+                });
+
+            var requests = new Stack<Request>(entryPoints);
+            var responses = new List<Response>();
+
+            while(requests.Count() > 0)
```

Comment:
```suggestion
            while(requests.Any())
```
shouldn't really matter here but still.